jib-core/src/main/java/com/google/cloud/tools/jib/filesystem/LockFile.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.filesystem;
+
+import java.io.Closeable;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.channels.FileLock;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/** Creates and deletes lock files. */
+public class LockFile implements Closeable {
+
+  private final Path lockFile;
+  private final FileLock lock;
+
+  private LockFile(Path lockFile, FileLock lock) {
+    this.lockFile = lockFile;
+    this.lock = lock;
+  }
+
+  /**
+   * Creates a lock file.
+   *
+   * @param lockFile the path of the lock file
+   * @return a new {@link LockFile} that can be released later
+   * @throws IOException if creating the lock file fails
+   */
+  public static LockFile lock(Path lockFile) throws IOException {
+    Files.createDirectories(lockFile.getParent());
+    while (true) {
+      try {
+        FileLock fileLock = new FileOutputStream(lockFile.toFile()).getChannel().tryLock();
```

Comment:
@dgageot is there a reason for manual polling instead of just using the blocking `FileChannel.lock()`. I can't think of why.

Reply:
Also this code catches the general `Exception`, so it doesn't honor thread interruptions. Seems too stringent blocking forever until it gets a lock. It should probably be interruptible?

Reply:
I tried using `lock()`, and it seems to throw `OverlappingFileLockException` instead of blocking if it's called in multiple threads in the same JVM.

Reply:
So maybe it needs to be synchronized on the same jvm?

Reply:
I mean using the `synchronized` keyword.

Reply:
Didn't seem to make a difference.

Reply:
When I used this code, it couldn't be called by multiple threads in the same JVM.
Why use `tryLock()` instead of `lock()` in a loop? I don't remember ;-).

Reply:
Ah yeah, no I misunderstood the problem. Still I think we can use some kind of semaphore here.

Reply:
Maybe a guava Striped<Semaphore> or something?
https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html

Reply:
I think `Striped` by be wrong, perhaps you can just use a `concurrentHashMap<Path, Lock>`.

Reply:
Sorry I'm just throwing things out here, but something along these lines.

Reply:
We'll try that for now